File: c/list-ops/src/list_ops.c
#include "list_ops.h"

#include <stdlib.h>

list_t *new_list(size_t length, list_element_t elements[]) {
    list_t *ret = (list_t *)malloc(sizeof(list_t));
    ret->length = length;
    ret->elements = (list_element_t *)malloc(sizeof(list_element_t) * length);

    for (size_t i = 0; i < length; ++i) {
        ret->elements[i] = elements[i];
    }

    return ret;
}

list_t *append_list(list_t *list1, list_t *list2) {
    list_t *ret = (list_t *)malloc(sizeof(list_t));
    ret->length = list1->length + list2->length;
    ret->elements = (list_element_t *)malloc(sizeof(list_element_t) * ret->length);

    size_t j = 0;
    for (size_t i = 0; i < list1->length; ++i) {
        ret->elements[j++] = list1->elements[i];
    }

    for (size_t i = 0; i < list2->length; ++i) {
        ret->elements[j++] = list2->elements[i];
    }

    return ret;
}

list_t *filter_list(list_t *list, bool (*filter)(list_element_t)) {
    int count = 0;
    for (size_t i = 0; i < list->length; ++i) {
        if (filter(list->elements[i])) {
            ++count;
        }
    }

    list_t *ret = (list_t *)malloc(sizeof(list_t));
    ret->length = count;
    ret->elements = (list_element_t *)malloc(sizeof(list_element_t) * count);

    size_t j = 0;
    for (size_t i = 0; i < list->length; ++i) {
        if (filter(list->elements[i])) {
            ret->elements[j++] = list->elements[i];
        }
    }

    return ret;
}

size_t length_list(list_t *list) {
    return list->length;
}

list_t *map_list(list_t *list, list_element_t (*map)(list_element_t)) {
    list_t *ret = (list_t *)malloc(sizeof(list_t));
    ret->length = list->length;
    ret->elements = (list_element_t *)malloc(sizeof(list_element_t) * list->length);

    for (size_t i = 0; i < list->length; ++i) {
        ret->elements[i] = map(list->elements[i]);
    }

    return ret;
}

// folds (reduces) the given list from the left with a function
list_element_t foldl_list(list_t *list, list_element_t initial, list_element_t (*foldl)(list_element_t, list_element_t)) {
    list_element_t acc = initial;
    for (size_t i = 0; i < list->length; ++i) {
        acc = foldl(acc, list->elements[i]);
    }

    return acc;
}

static list_element_t foldr_internal(int limit, int count, list_t *list, list_element_t initial,
                                     list_element_t (*foldr)(list_element_t, list_element_t)) {
    if (count == limit) {
        return initial;
    }

    list_element_t acc = foldr_internal(limit, count + 1, list, initial, foldr);
    return foldr(list->elements[count], acc);
}

list_element_t foldr_list(list_t *list, list_element_t initial, list_element_t (*foldr)(list_element_t, list_element_t)) {
    return foldr_internal(list->length, 0, list, initial, foldr);
}

list_t *reverse_list(list_t *list) {
    list_t *ret = (list_t *)malloc(sizeof(list_t));
    ret->length = list->length;
    ret->elements = (list_element_t *)malloc(sizeof(list_element_t) * list->length);

    for (size_t i = 0; i < list->length; ++i) {
        ret->elements[i] = list->elements[list->length - i - 1];
    }

    return ret;
}

void delete_list(list_t *list) {
    if (list == NULL) {
        return;
    }

    free(list->elements);
    free(list);
}

File: go/list-ops/list_ops.go
package listops

type IntList []int
type predFunc func(n int) bool
type binFunc func(x, y int) int
type unaryFunc func(n int) int

func (lst IntList) Filter(f predFunc) IntList {
	ret := make(IntList, 0)
	for _, n := range lst {
		if f(n) {
			ret = append(ret, n)
		}
	}

	return ret
}

func (lst IntList) Length() int {
	return len(lst)
}

func (lst IntList) Map(f unaryFunc) IntList {
	ret := make(IntList, 0)
	for _, n := range lst {
		ret = append(ret, f(n))
	}

	return ret
}

func (lst IntList) Reverse() IntList {
	ret := make(IntList, len(lst))
	i := 0
	j := len(lst) - 1

	for i < j {
		ret[i], ret[j] = lst[j], lst[i]
		i++
		j--
	}

	return ret
}

func (lst IntList) Append(other IntList) IntList {
	ret := make(IntList, len(lst))
	copy(ret, lst)

	for _, n := range other {
		ret = append(ret, n)
	}

	return ret
}

func (lst IntList) Concat(a []IntList) IntList {
	ret := make(IntList, len(lst))
	copy(ret, lst)

	for _, v := range a {
		for _, n := range v {
			ret = append(ret, n)
		}
	}

	return ret
}

func (lst IntList) Foldl(f binFunc, val int) int {
	ret := val

	for _, n := range lst {
		ret = f(ret, n)
	}

	return ret
}

func (lst IntList) Foldr(f binFunc, val int) int {
	if len(lst) == 0 {
		return val
	}

	v := lst[1:].Foldr(f, val)
	return f(lst[0], v)
}
